crates/xlineapi/src/command.rs

```diff
@@ -222,67 +230,136 @@ pub struct Command {
+/// ```
+macro_rules! match_all {
+    ($($cls:ident :: $tag:ident)&*) => {
+        |_x| $(matches!($cls::from(_x), $cls::$tag))&&*
+    };
+}
+pub(crate) use match_all; // used by lib.rs
 
-        // Lease leases request is conflict with Lease grant and revoke requests
-        if (this_req.is_lease_read_request() && other_req.is_lease_write_request())
-            || (this_req.is_lease_write_request() && other_req.is_lease_read_request())
-        {
-            return true;
+/// swapable match, returns a `Fn(x, y) -> bool` indicates the match result.
+///
+/// # Returns
+///
+/// Returns `Fn(x, y) -> true` if *x match Class1 and y match Class2*, or *x match Class2 and y match Class1*.
+macro_rules! swap_match {
+
+    ($($cls1:ident::$tag1:ident)&*, _) => {{
+        |_x, _| match_all!($($cls1::$tag1)&*)(_x)
+    }};
+    ($($cls1:ident :: $tag1:ident)&*, $($cls2:ident :: $tag2:ident)&*) => {
+        |_x, _y| {
+            (match_all!($($cls1::$tag1)&*)(_x) && match_all!($($cls2::$tag2)&*)(_y)) || (
+                match_all!($($cls2::$tag2)&*)(_x) && match_all!($($cls1::$tag1)&*)(_y)
+            )
         }
+    };
+}
 
-        if this_req.is_compaction_request() && other_req.is_compaction_request() {
-            return true;
+/// swapable map, to swappable match two `RequestWrapper`, extract the inner request
+/// and calculate the inners into an `Option<T>`.
+/// This can be only used in the same enum and different variant.
+///
+/// # Returns
+///
+/// `Fn(x, y) -> Option<bool>` indicates the swap_map result.
+macro_rules! swap_map {
+    ($cls1:ident :: $tag1:ident, $cls2:ident :: $tag2:ident, |$x:ident, $y:ident| $body:expr) => {
+        (|_self, _other| match (_self, _other) {
+            (&$cls1::$tag1(ref $x), &$cls2::$tag2(ref $y))
+            | (&$cls2::$tag2(ref $y), &$cls1::$tag1(ref $x)) => Some($body),
+            _ => None,
+        })
+    };
+}
+
+/// Conflict check, contains the whole `match` statement
+///
+/// # Returns
+///
+/// `Fn(x, y) -> Option<bool>` indicates the conflict result.
+macro_rules! is_conflict {
+    ($(($body:expr, $($cls1:ident :: $tag1:ident)&*, $($pat:tt)*)),*) => {
+        |_self, _other| match (_self, _other) {
+            $((x, y) if swap_match!($($cls1::$tag1)&*, $($pat)*)(x, y) => Some($body),)*
+            _ => None,
         }
+    };
+}
 
-        if (this_req.is_txn_request() && other_req.is_compaction_request())
-            || (this_req.is_compaction_request() && other_req.is_txn_request())
+impl ConflictCheck for Command {
+    #[inline]
+    fn is_conflict(&self, other: &Self) -> bool {
+        let t_req = &self.request;
+        let o_req = &other.request;
+        let mut cache_key: (u8, u8) = (t_req.into(), o_req.into());
+        if cache_key.0 > cache_key.1 {
+            cache_key = (cache_key.1, cache_key.0);
+        }
+        if let Some(res) = CONFLICT_RULES_CACHE
+            .read()
+            .ok()
+            .and_then(|c| c.get(&cache_key).cloned())
```

Comment:
Why do you need a cache? I believe the matches are already efficient. Or maybe use a thread local cache instead.

Reply:
thread_local used